frontend/src/store/modules/user.ts

```diff
@@ -51,32 +54,69 @@ export default {
     },
     setUserRoles: (state: any, roles: { [key: string]: any } | null) => {
       state.roles = roles;
+    },
+    setMetadataCallback: (
+      state: any,
+      callback: ((snapshot: firebase.database.DataSnapshot) => void) | null
+    ) => {
+      state.metadataCallback = callback;
+    },
+    setMetadataRef: (state: any, ref: any) => {
+      state.metadataRef = ref;
+    },
+    setTimestamp: (state: any, timestamp: number) => {
+      state.timestamp = timestamp;
     }
   },
   actions: {
     signOut() {
       firebase.auth().signOut();
     },
-    async handleUser(
+    async onAuthStateChanged(
+      { dispatch, commit }: ActionContext<any, any>,
+      user: firebase.User | null
+    ) {
+      commit("setCurrentUser", user);
+      dispatch("handleMetadata");
+    },
+    async handleInitialUserLoad(
       { commit, dispatch }: ActionContext<any, any>,
       user: firebase.User | null
     ) {
       commit("setCurrentUser", user);
-      if (user) {
-        await dispatch(
-          "updateUserRoles",
-          (await user.getIdTokenResult()).claims.roles
-        );
-      } else {
-        await dispatch("updateUserRoles", null);
-      }
+      await dispatch("updateUserRoles");
     },
-    updateUserRoles(
-      { commit, getters }: ActionContext<any, any>,
-      roles: { [key: string]: any } | null
+    async updateUserRoles(
+      { commit, getters, state }: ActionContext<any, any>,
+      { forceRefresh = false } = {}
     ) {
+      const roles = state.currentUser
+        ? (await state.currentUser.getIdTokenResult(forceRefresh)).claims.roles
+        : null;
       commit("setUserRoles", roles);
       getters.ability.update(defineAbilities());
+    },
+    handleMetadata({ state, dispatch, commit }: ActionContext<any, any>) {
+      if (state.metadataCallback && state.metadataRef) {
+        commit("setTimestamp", null);
+        state.metadataRef.off("value", state.metadataCallback);
+      }
+      if (!state.currentUser) {
+        return;
+      }
```

Comment:
Perhaps this means that the user has signed out. Don't we need to reset the abilities in this case?